kirkwood: add support for Cisco CVR328W-K9-CN

yangfl opened the pull request:
This patch adds support for Cisco CVR328W-K9-CN, sold as "2.4GHz Wireless VPN Router".

Specifications:
--------------

* SoC: Marvell 88F6282
* RAM: 256MB DDR3
* Flash: 128MB NAND flash (Hynix H27U1G8F2BTR-BC)
* WiFi 2.4GHz: Atheros AR9280 via Mini-PCIe (card with no label)
* Ethernet: Marvell 88E6095F-LG01 (8 FEs, 2 GbEs with 88E1114-NNC1, 1 uplink)
* LED: Power, Sys, USB, 3G, VPN, NMS, WLAN (via AR9280), 8 LANs (via 88E6095F), 2 WANs/ 1 LAN (via 88E1114)
* Buttons: Reset (GPIO 17)
* UART: Serial console (115200 8n1)
* USB: 2 x USB2

SPECIAL NOTES:
-------------

* At the time of this patch, the boot sequence is affected by the malfunctioned MANGLE_BOOTARGS. Either you should apply the fix for MANGLE_BOOTARGS, or disable it completely, or you will need to adjust the U-Boot environment variable manually, by deleting bootargs_root=root=/dev/mtdblock3 rw using `uboot_setenv` via console or `set` via U-Boot. Alternatively, you can `set bootargs console=ttyS0,115200` manually before U-Boot kernel booting.
* The bootloader blindly sets the creation time to -1 before deciding uImage size. If that check fails, kernel load size will be set to a fixed value (about 2M).

Installation:
------------

You can access the device console via the following methods.

* Go to web interface, navigate to System Management -> Remote Management -> SSH and enable SSH. Please note the SSH server only support
  diffie-hellman-group1-sha1, you won't be able to connect it with an up-to-
  date ssh client.
* Open the shell and use a UART2USB convert to gain TTY access (at JP1, from 1 to 5: VCC, TXD, RXD, NC, GND).

Please back up firmware if you want to go back to the original.

After you can control the device, flash the firmware as usual. Here are some hints for that.

!!! Please read SPECIAL NOTES before using any of the following methods.

Option 1 (via web interface):

  Flash factory.bin via OEM web interface.

Option 2 (via U-Boot):

  tftpboot $(loadaddr) <factory.bin>
  nand erase 0x2900000 $(filesize)
  nand write.e $(loadaddr) 0x2900000 $(filesize)
  # switch to second firmware
  set cur_booted_runtime_image 2
  saveenv
  # boot newly flash OpenWrt directly
  bootm $(loadaddr)

Option 3 (via OEM console):

  cd /tmp
  wget http://<your-computer-host>/<factory.bin>
  dd if=<factory.bin> of=/dev/mtdblock4 bs=64k
  # switch to second firmware
  uboot_setenv cur_booted_runtime_image 2

* Please do check mtdblock part num by /proc/mtd.

To go back to OEM firmware, flash cvr328w_rv315w_<ver>.img just like factory.bin or simply switch the dual boot partition.

Dual boot:
---------

The device supports dual boot. The patch for dual boot support of OpenWrt is still waiting for reviews, currently the patch use the second firmware partition only.

The firmware to boot is controlled by U-Boot env `cur_booted_runtime_image`. If `runtime_image<i>_try_num` reaches 3, bootloader will switch to another partition. If both firmwares fail to boot, bootloader will start a TFTP server waiting for new firmware (this feature is not tested). This process times out after ~30s.

If one firmware boots failed for 2 times, once the OEM firmware in another partition is booted, it will try to overwrite the former firmware with itself.

To switch between firmwares without console, power on the device, wait ~10s, then power it off, repeat these steps for 3 times.

MAC addresses:
-------------

The MAC addresses and other OEM configs are stored in NVRAM partition using U-Boot env format.

Signed-off-by: David Yang <mmyangfl@gmail.com>

Ansuel commented on include/image-commands.mk:
```diff
@@ -558,6 +558,7 @@ define Build/tplink-v2-image
 endef
 
 define Build/uImage
+	$(if $(UIMAGE_TIME),SOURCE_DATE_EPOCH="$(UIMAGE_TIME)") \
```
put this in another commit please

yangfl replied:
#11966

robimarko replied:
Not PR, just another commit.

Ansuel commented:
the 900 patch should be discussed in a separate pr or anyway introduced in a separate commit with some extra info

yangfl commented:
@Ansuel the 900 patch has been accepted by kernel, so we just put it here for the time being, and simply wait openwrt getting that newer version of kernel

Ansuel commented:
@yangfl A better approach is to put it in generic, in the backport directory and add a tag so we can drop once it's backported to a stable kernel.
Also doesn't look to be final version of the change by looking at the tags in the related patch.
Would be better to generate the patch from a linux clone. Do you have the commit of that patch merged upstream?

yangfl committed:
build: Allow specifying uImage time

Some U-Boot checks for a specified uImage time and refuses to boot if
mismatched. This patch fixes it by recognizing UIMAGE_TIME parameter.

Signed-off-by: David Yang <mmyangfl@gmail.com>

yangfl commented:
https://github.com/torvalds/linux/commit/d08cb25556773c65efb21761b75f92c804ad0975

yangfl committed:
kirkwood: backport "net: mv643xx_eth: support MII/GMII/RGMII modes for Kirkwood"

Signed-off-by: David Yang <mmyangfl@gmail.com>

yangfl committed:
kirkwood: add support for Cisco CVR328W-K9-CN

This patch adds support for Cisco CVR328W-K9-CN, sold as "2.4GHz Wireless
VPN Router".

Specifications:
--------------

* SoC: Marvell 88F6282
* RAM: 256MB DDR3
* Flash: 128MB NAND flash (Hynix H27U1G8F2BTR-BC)
* WiFi 2.4GHz: Atheros AR9280 via Mini-PCIe (card has no label)
* Ethernet: Marvell 88E6095F-LG01 (8 FEs, 2 GbEs with 88E1114-NNC1, 1
  uplink)
* LED: Power, Sys, USB, 3G, VPN, NMS, WLAN (via AR9280), 8 LANs (via
  88E6095F), 2 WANs/ 1 LAN (via 88E1114)
* Button: Reset (GPIO 17)
* UART: Serial console (115200 8n1)
* USB: 2 x USB2

SPECIAL NOTES:
-------------

* At the time of this patch, the boot sequence is affected by the
  malfunctioned MANGLE_BOOTARGS. You should apply the fix for
  MANGLE_BOOTARGS, or disable it completely. There is no way to workaround
  it within persistent U-Boot environment variables.

  Alternatively, you can `set bootargs console=ttyS0,115200` manually before
  U-Boot kernel booting.

* The bootloader blindly sets the creation time to -1 before deciding uImage
  size. If that check fails, uImage load size `load_kernel_size` will be set
  to a fixed value 0x280000. But if the kernel does fit that size, it can be
  booted successfully.

Installation:
------------

You can access the device console via the following methods.

* Go to web interface, navigate to System Management -> Remote Management ->
  SSH and enable SSH. Please note the SSH server only support
  diffie-hellman-group1-sha1, you won't be able to connect it with an up-to-
  date ssh client.

* Open the shell and use a UART2USB convert to gain TTY access (at JP1, from
  1 to 5: VCC, TXD, RXD, NC, GND).

Please back up firmware if you want to go back to the original.

After you can control the device, flash the firmware as usual. Here are some
hints for that.

!!! Please read SPECIAL NOTES before using any of the following methods.

Option 1 (via web interface):

  Flash factory.bin via OEM web interface.

Option 2 (via U-Boot):

  tftpboot $(loadaddr) <factory.bin>
  nand erase 0x2900000 $(filesize)
  nand write.e $(loadaddr) 0x2900000 $(filesize)
  # switch to second firmware
  set cur_booted_runtime_image 2
  set runtime_image2_try_num
  saveenv
  # boot newly flash OpenWrt directly
  bootm

Option 3 (via OEM console):

  cd /tmp
  wget http://<your-computer-host>/<factory.bin>
  # you need to download `dd` util (static busybox binary for example)
  dd if=<factory.bin> of=/dev/mtdblock4 bs=64k
  # switch to second firmware
  uboot_setenv cur_booted_runtime_image 2
  uboot_setenv runtime_image2_try_num

  * Please do check mtdblock part num by /proc/mtd.

To go back to OEM firmware, flash cvr328w_rv315w_<ver>.img just like
factory.bin or simply switch the dual boot partition.

Dual boot:
---------

The device supports dual boot. The patch for dual boot support of OpenWrt is
still waiting for reviews, currently the patch use the second firmware
partition only.

The firmware to boot is controlled by U-Boot env `cur_booted_runtime_image`.
If `runtime_image<i>_try_num` reaches 3, bootloader will switch to another
partition. If both firmwares fail to boot, bootloader will start a TFTP
server waiting for new firmware (this feature is not tested). This process
times out after ~30s.

If one firmware boots failed for 2 times, once the OEM firmware in another
partition is booted, it will try to overwrite the former firmware with
itself.

To switch between firmwares without console, power on the device, wait ~10s,
then power it off, repeat these steps for 3 times.

Multi WAN:
---------

The device is marketed with multi WAN support. This is purely implemented
using software by connecting all ethernet ports to one switch chip
(88E6095F). One port is marked with WAN1/LAN0, which is treated as LAN
interface in this patch. 2 LEDs, WAN1 and LAN0, correspond to that
interface, where WAN1 is STATUS[0] and LAN0 is STATUS[1]. LAN0/STATUS[1] is
always turned off, and switch between WAN1/LAN0 LEDs is not implemented in
this patch.

MAC addresses:
-------------

The MAC addresses and other OEM configs are stored in NVRAM partition using
U-Boot env format.

Signed-off-by: David Yang <mmyangfl@gmail.com>

sanderpqr commented:
@yangfl Now 6.6 is the default kernel, this commit has become a bit simpler?

yangfl commented:
I don't own that device anymore; feel free to reuse any code.
